modules/040-node-manager/hooks/instance_claim_controller.go

```diff
@@ -0,0 +1,308 @@
+	var ic d8v1alpha1.InstanceClaim
+
+	err := sdk.FromUnstructured(obj, &ic)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ic, nil
+}
+
+func instanceClaimController(input *go_hook.HookInput) error {
+	instanceClaims := make(map[string]*d8v1alpha1.InstanceClaim)
+	machines := make(map[string]*machineForInstanceClaim)
+	nodeGroups := make(map[string]*nodeGroupForInstanceClaim)
+
+	for _, i := range input.Snapshots["instances"] {
+		ins := i.(*d8v1alpha1.InstanceClaim)
+		instanceClaims[ins.GetName()] = ins
+	}
+
+	for _, m := range input.Snapshots["machines"] {
+		mc := m.(*machineForInstanceClaim)
+		machines[mc.Name] = mc
+	}
+
+	for _, m := range input.Snapshots["ngs"] {
+		ng := m.(*nodeGroupForInstanceClaim)
+		nodeGroups[ng.Name] = ng
+	}
+
+	// first, check mapping from machines to instance claims
+	// here we handle two cases:
+	//   1. if instance claim exists, then we have 2 subcases:
+	//  	a. first, we should compare machines fields and instance claim status and if we have changes path status
+	//		b. next, we should check metadata.deletionTimestamp for instance claim
+	//         if we have non-zero timestamp this means that client deletes instance claim, and we should delete machine.
+	//         Before deletion, we should check metadata.deletionTimestamp in the machine for prevent multiple deletion
+	//   2. if instance claim does not exist, create instance claim for machine
+	for name, machine := range machines {
+		if ic, ok := instanceClaims[name]; ok {
+			statusPatch := getInstanceClaimStatusPatch(ic, machine)
+			if len(statusPatch) > 0 {
+				patch := map[string]interface{}{
+					"status": statusPatch,
+				}
+				input.PatchCollector.MergePatch(patch, "deckhouse.io/v1alpha1", "InstanceClaim", "", ic.Name, object_patch.WithSubresource("/status"))
+			}
+
+			if ic.DeletionTimestamp != nil && !ic.DeletionTimestamp.IsZero() {
+				if machine.DeletionTimestamp == nil || machine.DeletionTimestamp.IsZero() {
+					// delete in background, because machine has finalizer
+					input.PatchCollector.Delete("machine.sapcloud.io/v1alpha1", "Machine", "d8-cloud-instance-manager", machine.Name, object_patch.InBackground())
+				}
+			}
+		} else {
+			ng, ok := nodeGroups[machine.NodeGroup]
+			if !ok {
+				return fmt.Errorf("NodeGroup %s not found", machine.NodeGroup)
+			}
+			newIc := newInstanceClaim(machine, ng)
+			input.PatchCollector.Create(newIc, object_patch.IgnoreIfExists())
+		}
+	}
+
+	// next, check mapping from instance claims to machines
+	// here we handle next cases:
+	//   1. if machine exists, then skip it, we handle it above
+	//   2. if machine does not exist:
+	//  	a. we should delete finalizers only if metadata.deletionTimestamp is non-zero
+	//		b. we should delete finalizers and delete instance claim if metadata.deletionTimestamp is zero
+	for _, ic := range instanceClaims {
+		if _, ok := machines[ic.GetName()]; !ok {
+			deleteFinalizersPatch := map[string]interface{}{
```

Comment:
This patch could be moved to the `var` block (it does not have variables)

Reply:
done